File: application/test/functional/gateway/AuthenticationIT.scala
package functional.gateway

import java.util.concurrent.TimeUnit

import akka.http.scaladsl.model.headers.{Authorization, BasicHttpCredentials, OAuth2BearerToken}
import api.rest.resources.AuthenticationResource
import com.github.tomakehurst.wiremock.WireMockServer
import com.github.tomakehurst.wiremock.client.WireMock._
import com.github.tomakehurst.wiremock.core.WireMockConfiguration.options
import common.{FunctionalSpec, WsConnectionException}
import common.api.EndpointsClient
import fixtures.EndpointFixtures
import play.api.test.Helpers._
import play.mvc.Http.Status

import scala.collection.immutable

class AuthenticationIT extends FunctionalSpec with EndpointsClient {
  private implicit val defaultTimeout: Int = 10000

  private val wiremockServer = new WireMockServer(options().dynamicPort())
  private var wiremockPort = 0
  private var authorizationServerUrl = ""

  override protected def beforeAll(): Unit = {
    wiremockServer.start()
    wiremockPort = wiremockServer.port()
    authorizationServerUrl = s"http://localhost:$wiremockPort/verify_token"
  }


  feature("Gateway Basic Authentication") {
    Given("endpoint who requires basic authentication")
    val path = "/basic-auth"
    val username = "cosmin"
    val password = "St0nGpa$$"
    val endpoint = EndpointFixtures.fromPath(path)
      .copy(authentication = AuthenticationResource.basic(username, password))
    createAndAssert(endpoint)

    scenario("Authorization header is sent with correct credentials") {
      When("Try to connect")
      val (in, out) = await(wsClient.connectWithHeaders(
        s"$wsHost$path",
        immutable.Seq(getBasicAuthorizationHeader(username, password))
      ))

      Then("Access granted")
      await(out.offer("Hi"), defaultTimeout, TimeUnit.MILLISECONDS)
    }

    scenario("Authorization header is sent with wrong credentials") {
      When("Try to connect")
      val caught = intercept[WsConnectionException] {
        await(wsClient.connectWithHeaders(
          s"$wsHost$path",
          immutable.Seq(getBasicAuthorizationHeader("s", "s"))
        ))
      }

      Then("Access denied")
      caught.getHttpCode() mustEqual Status.UNAUTHORIZED
    }
  }

  feature("Gateway Bearer Authentication") {
    scenario("Authorization header is sent with a good token") {
      Given("endpoint who requires bearer authentication")
      val path = "/bearer-auth"
      val endpoint = EndpointFixtures.fromPath(path)
        .copy(authentication = AuthenticationResource.bearer(authorizationServerUrl))
      createAndAssert(endpoint)
      wiremockServer.stubFor(
        post(urlPathEqualTo("/verify_token"))
          .withRequestBody(equalToJson("{\"access_token\": \"good\"}"))
          .willReturn(aResponse().withStatus(Status.OK))
      )
      When("Try to connect")
      val (in, out) = await(wsClient.connectWithHeaders(
        s"$wsHost$path",
        immutable.Seq(getBearerAuthorizationHeader("good"))
      ))

      Then("Access granted")
    }

    scenario("Authorization header is sent with a wrong token") {
      Given("endpoint who requires bearer authentication")
      val path = "/bearer-auth-access-denied"
      val endpoint = EndpointFixtures.fromPath(path)
        .copy(authentication = AuthenticationResource.bearer(authorizationServerUrl))
      createAndAssert(endpoint)

      wiremockServer.stubFor(
        post(urlPathEqualTo("/verify_token"))
          .withRequestBody(equalToJson("{\"access_token\": \"wrong\"}"))
          .willReturn(aResponse().withStatus(Status.FORBIDDEN))
      )

      When("Try to connect")
      val caught = intercept[WsConnectionException] {
        await(wsClient.connectWithHeaders(
          s"$wsHost$path",
          immutable.Seq(getBearerAuthorizationHeader("wrong"))
        ))
      }

      Then("Access denied")
      caught.getHttpCode() mustEqual Status.UNAUTHORIZED
    }
  }

  private def getBearerAuthorizationHeader(token: String) = {
    Authorization(OAuth2BearerToken(token))
  }

  private def getBasicAuthorizationHeader(username: String, password: String) = {
    Authorization(BasicHttpCredentials(username, password))
  }
}

File: ws-gateway-functional-tests/src/test/java/com/cosmin/wsgateway/tests/gateway/AuthenticationIT.java
package com.cosmin.wsgateway.tests.gateway;

import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.equalTo;
import static com.github.tomakehurst.wiremock.client.WireMock.post;
import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.cosmin.wsgateway.api.representation.AuthenticationRepresentation;
import com.cosmin.wsgateway.tests.BaseTestIT;
import com.cosmin.wsgateway.tests.Tags;
import com.cosmin.wsgateway.tests.client.WebSocketConnectionException;
import com.cosmin.wsgateway.tests.common.EndpointFixtures;
import com.cosmin.wsgateway.tests.common.JsonUtils;
import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;

@Tags.Gateway
public class AuthenticationIT extends BaseTestIT {
    public static WireMockServer wiremock = new WireMockServer(WireMockConfiguration.options().dynamicPort());
    private static String wiremockUrl;

    @BeforeAll
    static void beforeAll() {
        wiremock.start();
        wiremockUrl = String.format("http://localhost:%d", wiremock.port());
    }

    @AfterAll
    static void afterAll() {
        wiremock.stop();
    }

    @BeforeEach
    void setUp() {
        wiremock.resetAll();
    }

    @Test
    public void basicAuthenticationIsUsedAndChecked() {
        Given("endpoint with basic auth");
        var endpoint = EndpointFixtures.getRepresentation("/basic-auth");
        endpoint.setAuthentication(new AuthenticationRepresentation.Basic("username", "pass%6h"));
        endpointsClient.createAndAssert(endpoint);

        When("Connect with wrong auth");
        var thrown = assertThrows(WebSocketConnectionException.class, () -> {
            webSocketClient.connect("/basic-auth", Map.of("Authorization", "Basic wrongCredentials"));
        });
        Then("Connection is rejected with 401");
        assertEquals(HttpStatus.UNAUTHORIZED, thrown.getHttpStatus());

        When("Connect with good auth");
        var goodToken = Base64.getEncoder().encodeToString("username:pass%6h".getBytes(StandardCharsets.UTF_8));
        Then("Connection is accepted");
        webSocketClient.connect("/basic-auth", Map.of("Authorization", "Basic " + goodToken));
    }

    @Test
    public void bearerAuthenticationIsUsedAndChecked() {
        Given("endpoint with bearer auth and an authorization server");
        var endpoint = EndpointFixtures.getRepresentation("/bearer-auth");
        endpoint.setAuthentication(new AuthenticationRepresentation.Bearer(wiremockUrl + "/verify_token"));
        endpointsClient.createAndAssert(endpoint);

        var validTokenPayload = JsonUtils.toJson(Map.of("access_token", "validToken"));
        var invalidTokenPayload = JsonUtils.toJson(Map.of("access_token", "invalidToken"));

        wiremock.stubFor(post(urlPathEqualTo("/verify_token"))
                .withRequestBody(equalTo(validTokenPayload))
                .willReturn(aResponse()
                        .withHeader("Content-Type", "application/json")
                        .withBody("{}")));
        wiremock.stubFor(post(urlPathEqualTo("/verify_token"))
                .withRequestBody(equalTo(invalidTokenPayload))
                .willReturn(aResponse()
                        .withStatus(HttpStatus.FORBIDDEN.value())
                        .withHeader("Content-Type", "application/json")
                        .withBody("{}")));


        When("Connect with wrong auth token");
        var thrown = assertThrows(WebSocketConnectionException.class, () -> {
            webSocketClient.connect(endpoint.getPath(), Map.of("Authorization", "Bearer invalidToken"));
        });
        Then("Connection is rejected with 401");
        assertEquals(HttpStatus.UNAUTHORIZED, thrown.getHttpStatus());

        When("Connect with good auth token");
        webSocketClient.connect(endpoint.getPath(), Map.of("Authorization", "Bearer validToken"));
        Then("Connection is accepted");
    }
}
